### ISSUE TITLE
Payment pointers and internationalized domain names

### ISSUE THREAD

**Issue by dappelt** [interledger/rfcs#375]
I created this issue to discuss potential problems and solution of internationalized domain names in payment pointers. [Payment pointers](https://github.com/interledger/rfcs/blob/5641d91e806a8c3e27d97b91c76cacd13a87444b/0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md#appendix-a-payment-pointer) are introduced as part of SPSPv2. They support [internationalized domain names](https://en.wikipedia.org/wiki/Internationalized_domain_name) the same way URLs do, which makes payment pointers susceptible to [homograph attacks](https://en.wikipedia.org/wiki/IDN_homograph_attack).

To give a concrete example, let's assume alice is regularly paying to `$bob.example.com`. The invoices she is paying look something like `$bob.example.com/invoice/12345/whatever`
Mallory is aware of this and sends a forged payment pointer to Alice that looks like `$bob.example.com∕invoice∕12345∕.evil.com`. Note that there are no backslashes in that payment pointer, but the unicode char \u2215 (`∕ `), so the host is `evil.com`. Since both payment pointers look identical when displayed, Alice might be tricked in querying evil.com instead of example.com.

**Comment by michielbdejong** [comment]
I vote for restricting to ASCII. That way, you can still phish with `$exampIe.com/bob` which will look like $exampIe.com/bob, but at least that's easier for mainstream users to comprehend than weird unicode characters.

Also, as I said on the call, I have never seen anybody use international characters in their email address either, and even for instance Japanese native speakers will probably be able to use just ASCII characters without much inconvenience, when picking a payment pointer.

**Comment by adrianhopebailie** [comment]
> I vote for restricting to ASCII.

I vote for leaving international support but mentioning the risks in the "Security Considerations" section of the specification.

1. It is much simpler and easier for implementors who simply re-use whatever libraries and logic they already have for URL parsing to parse these. There is no character manipulation required, simply drop the leading `$`, split the pointer on the first `/` and construct a URL with the pieces.
2. How does a service that already hosts other services at a domain that uses non-latin characters create payment pointers for their users? Do we expect them to register a new domain just to host payment services.
3. By restricting to ASCII to prevent phishing we signal to implementors that phishing has been "solved". A half solution is worse than no solution. Nobody should ever trust a payment pointer on it's own. All setup protocols MUST provide a way for the sender to verify that they are communicating with the end receiver they expect. The pointer is just what it says it is, a pointer.

> Also, as I said on the call, I have never seen anybody use international characters in their email address either

This is a very Westernised perspective I'm afraid. See the section on adoption here: https://en.wikipedia.org/wiki/International_email

**Comment by emschwartz** [comment]
I agree that we should allow Unicode characters but I think this is unrealistic:

> All setup protocols MUST provide a way for the sender to verify that they are communicating with the end receiver they expect.

This has to rely on the TLS certificate, because anything a human could verify in the SPSP details could be copied by the attacker, but making sure people are talking to the right cert is notoriously challenging. I don't see a better solution for this than what browsers like Chrome do today

**Comment by dappelt** [comment]
> This has to rely on the TLS certificate

How exactly should this work? The original host and the spoofed host could both have valid certs (although for different domains of course). Sure, the wallet software could warn the user if he is paying to a cert for the first time, but how effective is that with non-tech-savvy users?

Also, using certs to identify receivers collides somewhat with this statement from the SPSPv2 spec:
> Adding this subdomain allows you to use $example.com as your payment pointer, even if the actual example.com is running a website via a CDN like github pages.

Multiples receivers hosted at such a CDN presumably share the same cert. So the sender cannot distinguish different receivers at the same CDN based on the cert. Combining this with the fact that there are two different ways of specifying payment pointers (i.e. as subdomain or in the path) opens another door for spoofing attacks.

For example, Bob might serve his SPSP details from `$bob.example.com`. Now if Mallory wants to spoof Bob, she could register `$example.com/bob`. Just from looking at the payment identifier you can't tell which one is the real Bob. TLS certs also don't help to distinguish the two payment pointers, since both are served from `example.com` using the same cert.

**Comment by LouSparfell** [comment]
Why don't you use IDNA encoding like browsers?

**Comment by adrianhopebailie** [comment]
@LouSparfell ca you expand on how you see that working?

If the payment pointer uses Internationalized Domain Name chars what URL would it resolve to?

**Comment by LouSparfell** [comment]
I'm not an expert but in browsers, but you can't really use unicode. It's converted with Punycode in ascii (it always begin with xn--) Else you won't be able to avoid homograph attacks and it can be stealth and painful.

For example with Payment Pointers :
$rïpple.com/invoice/12345/whatever would be converted in $xn--rpple-cta.com/invoice/12345/whatever
$example.com/invoice/12345/rïpple would be converted in $example.com/invoice/12345/xn--rpple-cta

**Comment by LouSparfell** [comment]
http://www.faqs.org/rfcs/rfc3492.html

**Comment by adrianhopebailie** [comment]
What @LouSparfell points out though, which is correct and was also mentioned to me in person by @sappenin, is that payment pointers don't natively support internationalized domain names.

The payment pointers spec inherits directly from the specification for https URLs which don't support anything but ASCII since DNS only supports ASCII.

The IDNA spec(s) define a way to turn an internationalized domain into a valid domain so it can be used in DNS and to get the resource at the given normalized URL.

It will be up to wallets and other user facing applications to decide if they will support internationalized payment pointers and will convert these (behind the scenes) into the appropriate ASCII form before starting to setup a payment.

Example: If I give you the payment pointer `$ärger.de` then your client software must first convert that to  `xn--rger-koa.de` before resolving it to the receiver endpoint URL `https://xn--rger-koa.de/.well-known/pay` and then doing an HTTP GET to that URL.

The client software could legitimately reject that pointer OR give an extra warning to the user that it is being converted from an internationalized form. I think this is out of scope for this spec so I suggest we can close this issue.

**Comment by dappelt** [comment]
> The payment pointers spec inherits directly from the specification for https URLs which don't support anything but ASCII since DNS only supports ASCII.

Ok, then the spec should state explicitly that the character set for payment pointers is limited to ASCII. The spec should also be specific about the relation of payment pointers to Punycode. In other words, are implementations supposed to transcode Punycode into internationalized payment pointers? For the time being, I recommend not to transcode Punycode until we figure out how to better protect users from phishing.

> It will be up to wallets and other user-facing applications to decide if they will support internationalized payment pointers and will convert these (behind the scenes) into the appropriate ASCII form before starting to setup a payment.

Before pushing the problem onto another layer we should think through how wallet implementations are supposed to handle internationalized payment pointers. The goal of this exercise is to assess whether wallets could implement suitable phishing countermeasures or if we need to make changes to SPSP/payment pointers to prevent phishing.

Let's recap.

## Problem Statement

How can a user or an implementation verify that an (internationalized) payment pointer is resolving to the SPSP endpoint of the intended recipient? Phishing techniques may trick a user into paying a familiar looking payment pointer, yet unintended recipient. Examples are [homograph attacks](https://en.wikipedia.org/wiki/IDN_homograph_attack) and [typosquatting](https://en.wikipedia.org/wiki/Typosquatting).

I think phishing will become a bigger problem for payment pointers+SPSP than it is for websites. In comparison, phishing websites require some effort to look convincing while spoofed SPSP endpoints can more or less simply copy the data from the intended SPSP endpoint. Phishing websites also need to come up with a believable reason why a user should enter their password and a second factor, which makes careful users suspicious.

## Potential Solutions

**EV Certificates** (proposed by @adrianhopebailie). Upon receiving a given payment pointer for the first time, a wallet could prompt the user to review the certificate details of the SPSP server. Subsequent payments to the same pointer do not have to be reviewed again by the user.

I don't think EV certificates will solve the problem. It is easy to obtain [fake EV certificates](https://stripe.ian.sh/) and they don't help [against](https://www.bleepingcomputer.com/news/security/extended-validation-ev-certificates-abused-to-create-insanely-believable-phishing-sites/) [phishing](https://en.wikipedia.org/wiki/Extended_Validation_Certificate#Domain_validated_certificates_were_created_by_CAs_in_the_first_place). Also, making EV certificates a requirement to run an SPSP server would hamper adoption, since they are expensive to obtain.

Another drawback is that a certificate is tied to the domain (i.e. SPSP server) and not to a particular SPSP endpoint. For example, if `example.com` is an SPSP hosting providing, the legitimate receiver `$bob.example.com` will have the same certificate as a spoofed receiver `$b0b.example.com`. And there are other problems with using certificates for this purpose. For example, if the certificate is renewed the wallet implementation detects that the certificate changed and gives a false alarm about a phishing attempt.

**Warn/reject if pointer contains Punycode.**
> The client software could legitimately reject that (internationalized) pointer OR give an extra warning to the user that it is being converted from an internationalized form.

Showing a warning before displaying internationalized payment pointers does not help users of non-latin alphabets. Such payment pointers, no difference if legitimate or not, would always prompt a warning and quickly become ignored by users. In addition, users that send mostly to ASCII pointers would get irritated by such a warning when occasionally paying to a legitimate internationalized pointer. Either we support internationalized pointers or we don't. Something in the middle does help neither user group.

**Picture in SPSP response.**
For each payment pointer, the user could select a picture that must be returned in the SPSP response. A spoofed SPSP endpoint does not know the picture and, hence, a user could recognize phishing attempts (essentially what okta.com does to prevent phishing).

However, I discarded this idea again since a user would need to remember too many distinct pictures (one for each pointer).

**Secure payment setup as part of SPSP.**
I am currently looking into how a payment pointer can be tied to an SPSP endpoint on the protocol level. SPSP could be extended so that subsequent calls to the same payment pointer verify that both sides have knowledge of a secret. Such a secret would need to be established the first time a given wallet and SPSP endpoint interact. Phishing attempts fail since the spoofed SPSP endpoint does not have the secret. The name of the protocol I am looking into is [secure remote password](https://en.wikipedia.org/wiki/Secure_remote_password_protocol).

## Conclusion

I don't have a definitive answer yet how to make payment pointers less susceptible to phishing. For the time being, I propose the spec recommends not to transcode Punycode.

**Comment by adrianhopebailie** [comment]
> the spec should state explicitly that the character set for payment pointers is limited to ASCII. 

I think that is implied by the fact that it is effectively a limited profile of https URL but will see if there's somewhere to add it that makes sense.

> I am currently looking into how a payment pointer can be tied to an SPSP endpoint on the protocol level.

I don't think this is the correct way to solve this problem. This assumes that some security will have been performed by the time the sender has resolved the receiver endpoint URL or that binding the pointer to the URL is a useful security property which is not true.

A receiver endpoint URL MUST be assumed to be malicious until proven otherwise. The solutions for verifying that the receiver is who the sender expects belong in SPSP or whatever setup protocol is used.

> I don't have a definitive answer yet how to make payment pointers less susceptible to phishing. For the time being, I propose the spec recommends not to transcode Punycode.

I think this is a useful summary of the issues but in reality these are all issues with SPSP not with Payment Pointers. 

The Payment Pointers spec is a very simple mapping from one identifier format to another, I think highlighting these issues is useful but it's not for this spec to try and solve them.

I don't agree that explicitly recommending not to transcode Punycode is a solution.

I would recommend raising this issue against SPSP and tracking it there.

**Comment by adrianhopebailie** [comment]
@dappelt I have logged an issue against SPSP at #392.

I think the only outstanding question is whether it should be allowed to resolve an internationalized Payment Pointer to a PunyCode URL. i.e. Does `$ärger.de` get resolved to `https://xn--rger-koa.de/.well-known/pay`?

In my opinion that is NOT supported by the current spec which expects the payment pointer to already be in ASCII (it must be valid URL parts).

If a user provides a payment pointer like `$ärger.de` then the spec defines no way to resolve that. The payment pointer would have to already be in the ASCII form: `$xn--rger-koa.de`.

A new specification (like the set of IDNA specs) that defines how to convert `$ärger.de` to `$xn--rger-koa.de` needs to be provided otherwise this behaviour is "unspecified".

For that reason I recommend we close this issue and continue pursuing a solution to #392

**Comment by emschwartz** [comment]
Regardless of what the spec says, I think most implementations of Payment Pointer resolvers are going to intentionally or accidentally support IDNs. For example, @sharafian's [implementation](https://github.com/sharafian/ilp-protocol-spsp/blob/master/index.js#L22-L24) will accept IDNs because the HTTP client library used automatically resolves them.

I think there are issues here that need to be discussed in the context of Payment Pointers, rather than just SPSP. For example, what happens if you have a vanity domain (like my $evan.cash) that redirects to another SPSP server and someone modifies the DNS record or the domain to redirect to? My SPSP endpoint is hosted by someone else entirely but it's probably my payment pointer domain that you'd want to go after if you're trying to steal money meant for me.

The way I'd try to break it down is based on where you get the payment pointer from and what the likely issues are in each case:

| Source | Potential Problems |
|---|---|
| Anywhere | Someone changes the DNS record or redirect target to point to their SPSP endpoint |
| Spoken aloud and transcribed | If you mistype it by accident, you may send money to the wrong person, but most people will probably have easy-to-type pointers for this reason |
| Web Payments API and HTTP-ILP | It'll be communicated in the background so homographic attacks won't come into play (are there any other issues that might come up?) |
| Email, chat or text messages, and Web pages | This is probably the most dangerous one, and where phishing is likely to be a problem, because your messaging client or web browser might be configured to recognize payment pointers and have your payment software automatically handle them |

- Should we recommend against highlighting payment pointers like email addresses? Would that type of recommendation ever be followed?
- Should we recommend against ever putting payment pointers in digital communications? (for example, shops should never send you payment pointer invoices that you're expected to click on)
- Should we recommend that punycode be copied if you copy a payment pointer from a browser (like some browsers do with copying URLs from the address bar)?
- Should payment client software show you punycode by default when confirming the payment?

**Comment by adrianhopebailie** [comment]
> Regardless of what the spec says, I think most implementations of Payment Pointer resolvers are going to intentionally or accidentally support IDNs.

Okay, what would you suggest changes in the spec which already has a pretty comprehensive Security Considerations section: https://interledger.org/rfcs/0026-payment-pointers/#security-considerations

> Should we recommend against highlighting payment pointers like email addresses? Would that type of recommendation ever be followed?

I don't follow what this is suggesting.

> Should we recommend against ever putting payment pointers in digital communications? (for example, shops should never send you payment pointer invoices that you're expected to click on)

Why? Obviously that will be ignored. The spec is for implementors of clients and servers not users.

> Should we recommend that punycode be copied if you copy a payment pointer from a browser (like some browsers do with copying URLs from the address bar)?

Why is this in scope of the spec that defines the resolution protocol? Who is this recommendation for? 

> Should payment client software show you punycode by default when confirming the payment?

Payment client software should adhere to the spec and NOT depend on the pointers as a way to identify the receiver. It doesn't matter if it shows punycode or not. All that will do is provide for a surprising and therefore poor UX.

The spec says:

> Sending clients MUST provide a way for a sender, after providing a pointer to the client, to verify that the entity they are sending to (or at least the entity hosting the payment setup protocol receiver endpoint) is who they intended otherwise senders can easily be fooled into sending money to the wrong receiver. 

**IMPORTANT:** Payment Pointers are not like email addresses. There is no standard mechanism for my email client to get information about the recipient of an email before I send it. Senders are not sending a payment to a payment pointer, they are using that pointer to provide a root for the client to discover the receiver. Only after that is done should the client attempt to verify that the receiver is who the sender intends.

**Comment by sharafian** [comment]
> Sending clients MUST provide a way for a sender, after providing a pointer to the client, to verify that the entity they are sending to (or at least the entity hosting the payment setup protocol receiver endpoint) is who they intended otherwise senders can easily be fooled into sending money to the wrong receiver.

I don't think this is a very helpful thing to mandate in the specification if it doesn't give a specific way to implement it.

When you're sending money to someone you don't yet know, I personally agree with Enrique's sentiment here: https://github.com/interledger/rfcs/issues/392#issuecomment-365630717 . PKI schemes just get compromised and/or ignored by the user. Whenever you can, you should send a small amount and wait for a small service to be rendered, rather than paying an entire price upfront. This massively reduces the incentive for the attacker.

For the peer-to-peer payments case, I think it's the responsibility lies in the site that transmits the payment pointer, not in the wallet. When I receive a chat message on Facebook, it's Facebook that's vouching for my friend's identity. So if I receive a payment pointer over FB messenger, it's facebook's security that gives me trust in the payment pointer, not the pointer itself. The payment pointer is always as safe as the channel it's transmitted over.

**Comment by adrianhopebailie** [comment]
> I don't think this is a very helpful thing to mandate in the specification if it doesn't give a specific way to implement it.

So are you suggesting it should come out or that we use this simple spec as a means to solve probably the biggest challenge of our time, hardened digital identity.

> When you're sending money to someone you don't yet know, I personally agree with Enrique's sentiment here: #392 (comment) . PKI schemes just get compromised and/or ignored by the user. Whenever you can, you should send a small amount and wait for a small service to be rendered, rather than paying an entire price upfront. This massively reduces the incentive for the attacker.

None of this says anything about what should or could change in the payment pointers spec. These are solutions to the problem that the spec says exists but is out of scope for the spec itself to prescribe a solution.

> For the peer-to-peer payments case, I think it's the responsibility lies in the site that transmits the payment pointer, not in the wallet. 

Now we are just getting ridiculous. Facebook vouches for your friends identity in so far as they can. They don't vouch for the data he types in via messenger. If you get a payment pointer over FB all you should feel confident in is that it comes from the person that Facebook claims it comes from, not that it is the correct payment pointer to pay them.

> The payment pointer is always as safe as the channel it's transmitted over.

How a payment pointer is transmitted is entirely irrelevant. All payment pointers should be treated as untrusted unless they can be proven otherwise. You are confusing the integrity of the payment pointer with its validity for an expected purpose.

This is the whole point in separating what can be solved by payment pointers and what can be solved in the actual exchange of data between parties via SPSP.

The only way to guarantee that a payment pointer is correct is to send some money to the account it points to and then find out if that arrived. Maybe a slight tweak of @earizon 's auth protocol proposal where a micropayment is sent to a pointer and the response must contain a challenge that only the true recipient knows? (Some thing like the PIN agreement protocol in BlueTooth Pairing)

The next best thing is for the account owner to be able to sign a challenge you issue and so prove they control a key that also happens to be used to sign something in the SPSP response.

**Comment by dappelt** [comment]
>> The payment pointer is always as safe as the channel it's transmitted over.

 > How a payment pointer is transmitted is entirely irrelevant. All payment pointers should be treated as untrusted unless they can be proven otherwise. You are confusing the integrity of the payment pointer with its validity for an expected purpose.

@adrianhopebailie There is some confusion about the threat we are discussing. Let me restate it: An attacker creates a payment pointer that is visually indistinguishable from a payment pointer the victim **has already trust in**. By impersonating the owner of the payment pointer, the attacker tricks the victim to pay the look-a-like payment pointer, which resolves to the attacker's SPSP endpoint.

There are many reasons why the victim may already trust a pointer (e.g. the owner is supposedly a personal friend, a webshop where the victim makes frequent purchases, your bank etc.). 

You can compare this to phishing attacks that attempt to, e.g., steal [crypto wallet credentials](http://blog.talosintelligence.com/2018/02/coinhoarder.html#more). In this scenario, the attacker registers a domain name similar to a well-known crypto wallet provider and serves a page that looks like the real page. If the victim does not notice, it may give away wallet credentials.

**Comment by adrianhopebailie** [comment]
> An attacker creates a payment pointer that is visually indistinguishable from a payment pointer the victim has already trust in.

This is a mistake. A user shouldn't **ever** have trust in a payment pointer.

That is why I don't think you can compare payment pointers to phishing attacks.

A payment pointer is just a way to discover an endpoint and only after doing this should the sender attempt to validate that the **endpoint** is the one they expect (no matter what the payment pointer says). i.e. They should only pay a receiver discovered via a payment pointer after the setup protocol has allowed them to _verify that the receiver is who they expect_.

As an example, I would design the UI so that as soon as the user has entered the Payment Pointer and clicks "Lookup", the Payment Pointer disappears and is replaced immediately with some Receiver Info from the SPSP response.

This is the difference between payment pointers and URLs and why I think phishing is a bad analogy.

When a user puts a URL into their browser there is no protocol to verify the entity behind the URL other than EV certificates which we know to be ignored by users because they just look at the URL and are distracted by the content of the HTTP response.

When HTTP was designed the idea of needing to challenge/interrogate the server to prove who they are hadn't occurred to anyone yet so it is a bolt on for HTTPS and isn't a very good solution. We can do better with SPSP but the solution is in SPSP not Payment Pointers.

**Comment by emschwartz** [comment]
> This is a mistake. A user shouldn't ever have trust in a payment pointer.

That doesn't seem very realistic to me. If you tell me your Payment Pointer is `$adrian.hopebailie.com` and I want to send you money for the first time, what could I use aside from your word to tell if I'm sending money to the right place? 

> I would design the UI so that as soon as the user has entered the Payment Pointer and clicks "Lookup", the Payment Pointer disappears and is replaced immediately with some Receiver Info from the SPSP response.

If someone is intentionally trying to steal money, they'll make the data that comes back from the fake pointer exactly the same as what comes back from the real one, so that doesn't really help.

Setup protocols or wallet UIs _might_ be able to prevent you from paying the wrong person the second time you're using the same payment pointer, but I don't think there's a whole lot you can do on the first occasion.

**Comment by adrianhopebailie** [comment]
> what could I use aside from your word to tell if I'm sending money to the right place?

An assurance from the setup protocol that I am speaking to the correct endpoint.

> If someone is intentionally trying to steal money, they'll make the data that comes back from the fake pointer exactly the same as what comes back from the real one, so that doesn't really help.

Which is why I keep saying we need a way to have a challenge or something in SPSP so that there is a second factor the sender can use to ensure the receiver is not being spoofed. This can't be fixed in Payment Pointers, they are static and simply resolve to a URL. It can be fixed in SPSP which is a dynamic so the interaction between the sender and receiver can include steps to verify the identity of the receiver.

**Comment by sharafian** [comment]
> It can be fixed in SPSP which is a dynamic so the interaction between the sender and receiver can include steps to verify the identity of the receiver.

@adrianhopebailie: That puts a massive dependency on SPSP as it requires it to pull in some kind of identity system that ties the endpoint to real world identities. This isn't even a solved problem, except when you have a central database like a social media site.

If the channel through which the payment pointer is transmitted is responsible for identity (i.e. facebook messenger) then we know that the correct person is telling us who to pay. Whether the pointer goes to "their" wallet doesn't matter, because it's their decision as to where they want to receive money.

> An attacker creates a payment pointer that is visually indistinguishable from a payment pointer the victim has already trust in.

@dappelt: I think the best solutions in this case so far have been the deterministic image generated from the pointer (like what ethereum wallets do) and/or having wallets save known payment pointers so you'll see that the address is unknown if you're sending to an indistiguishable-but-different address. Wouldn't those be enough to unblock us for the time being?

**Comment by adrianhopebailie** [comment]
> That puts a massive dependency on SPSP

Yes. That is where the validation should happen not during the resolution of the endpoint from the payment pointer.

> it requires it to pull in some kind of identity system that ties the endpoint to real world identities

Not necessarily. Most attacks can be thwarted simply by proving the SPSP response is not a replay (i.e. copied from the legitimate SPSP endpoint)

> If the channel through which the payment pointer is transmitted is responsible for identity (i.e. facebook messenger) then we know that the correct person is telling us who to pay.

Sure but that is a very different situation to the one @dappelt and @emschwartz are talking about. If the pointer is delivered over a secure channel then it's not likely to be subject to homographic attacks either is it.

**Comment by sharafian** [comment]
> > it requires it to pull in some kind of identity system that ties the endpoint to real world identities

> Not necessarily. Most attacks can be thwarted simply by proving the SPSP response is not a replay (i.e. copied from the legitimate SPSP endpoint)

How do you authenticate that without having an identity system in place, or some kind of PKI?

**Comment by adrianhopebailie** [comment]
Firstly, "some kind of PKI" is not real-world identities. Binding an digital identity to a real-person is a very different and very difficult thing to do.

Once you have a public key you are happy belongs to me you can visit any SPSP endpoint claiming to be me and verify it does indeed belong to me. I persuaded you once to make the association in your own wallet that that public key is mine. That is not complex PKI, it's very basic trusted key exchange.

Second, here's a simple proposed solution. It solves for a specific use case. Whenever my wallet is online it synchronizes with my SPSP receiver and loads 100 random words into a word list. The list rotates according to a schedule know to both the wallet and the receiver.

When I want to give my Payment Pointer to someone I give them one of these words too. They present a challenge in the SPSP request that is a hash of a nonce and the word. The SPSP receiver can very quickly figure out which word was used and return this in the response allowing the sender to verify the SPSP endpoint.

There are solutions. Just because they take some work to find doesn't mean we should somehow change the Payment Pointers spec
